README.md
<h1 align="center">
  TuanisSidebar
</h1>

<h2 align="center">Flutter package to create sidebars specially for web, but can be integrated as a drawer in a mobile app. Allows nested items.</h2>


|Desktop|Mobile|
|---|---|
|![screenshot](https://github.com/ebreness/tuanis_sidebar/blob/main/doc/desktop.gif?raw=true "Desktop")|![screenshot](https://github.com/ebreness/tuanis_sidebar/blob/main/doc/mobile.gif?raw=true "Mobile")|

## Basic Example
```dart
return Scaffold(
  body: Row(
    crossAxisAlignment: CrossAxisAlignment.stretch,
    children: [
      TuanisSidebar(
        selectedColor: Colors.white,
        selectedTileColor: Colors.blue,
        selectedItemId: 'settings',
        sections: const [
          TuanisSidebarSection(
            title: Text(
              'MAIN',
              style: TextStyle(letterSpacing: 1.2),
            ),
            items: [
              TuanisSidebarItem(
                id: 'dashboard',
                tile: ListTile(
                  leading: Icon(Icons.dashboard),
                  title: Text('Dashboard'),
                ),
                items: [
                  TuanisSidebarItem(
                    id: 'dashboard_1',
                    tile: ListTile(
                      leading: Icon(Icons.abc),
                      title: Text('Child of Dashboard'),
                    ),
                  )
                ],
              ),
              TuanisSidebarItem(
                id: 'settings',
                tile: ListTile(
                  title: Text('Settings'),
                  leading: Icon(Icons.settings),
                ),
              )
            ],
          ),
          TuanisSidebarSection(
            borderColor: Color(0xffcbd5e1),
            items: [
              TuanisSidebarItem(
                id: 'logout',
                tile: ListTile(
                  leading: Icon(Icons.logout),
                  title: Text('Logout'),
                ),
              )
            ],
          ),
        ],
      ),
      Expanded(
        child: Container(
          padding: const EdgeInsets.all(50),
          color: Colors.white,
          child: const Text('App body here'),
        ),
      )
    ],
  ),
);
```


CHANGELOG.md
## 0.0.1
- Initial release - basic sidebar

## 0.0.2
- added vertical scroll bars
- ensure all item ids are unique
- ability to choose the colors for everything

## 0.1.0
- nested items (N levels)
- each item now receives a ListTile component to facilite color customizations
- global (apply to all items) color configuration for states like hover, focus, selected.

## 0.2.0
- items can be divided in sections

## 0.3.0
- collapse/expand animation

## 0.3.1
- selectedColor is now required. This fixes https://github.com/ebreness/tuanis_sidebar/issues/1
- some typos were fixed

## 0.3.2
- removed padding around sections

## 0.3.3
- removed margin around sections

lib/src/item.dart
import 'package:flutter/material.dart';
import 'package:tuanis_sidebar/src/core/ext.dart';
import 'package:tuanis_sidebar/src/core/state.dart';

/// Represents an item in the sidebar
///
/// [id] The item identifier. Used to know if this is the active item
/// [tile] A ListTile component. The properties of this component will override any property from the Sidebar component
/// [expandIcon] The icon to show when the item has subitems. The default value is `arrow_forward_ios`
/// [isSelected] If true, then the item is shown as selected
/// [items] The sub-items
class TuanisSidebarItem extends StatefulWidget {
  /// The item identifier. Used to know if this is the active item
  final String id;

  /// A ListTile component. The properties of this component will override any property from the Sidebar component
  /// If the tile Icon is not defined, the "abc" icon will be shown in collapse view
  final ListTile tile;

  /// The icon to show when the item has subitems. The default value is `arrow_forward_ios`
  final Icon? expandIcon;

  /// If true, then the item is shown as selected
  final bool isSelected;

  /// The sub-items
  final List<TuanisSidebarItem> items;

  /// Message to display when the item is hovered
  final Text? tooltip;

  const TuanisSidebarItem({
    super.key,
    required this.id,
    required this.tile,
    this.expandIcon = const Icon(
      Icons.arrow_forward_ios,
      size: 16,
    ),
    this.isSelected = false,
    this.items = const [],
    this.tooltip,
  });

  @override
  State<StatefulWidget> createState() => _TuanisSidebarItem();

  TuanisSidebarItem copyWith(
      {required ListTile extTile, required bool extIsSelected}) {
    return TuanisSidebarItem(
      id: id,
      tile: extTile,
      expandIcon: expandIcon,
      isSelected: extIsSelected,
      items: items,
      tooltip: tooltip,
    );
  }
}

class _TuanisSidebarItem extends State<TuanisSidebarItem> {
  @override
  void initState() {
    super.initState();
    assert(widget.tile.leading == null || widget.tile.leading is Icon);
  }

  @override
  Widget build(BuildContext context) {
    final isSidebarCollapse = SidebarState.of(context)?.isCollapse ?? false;

    Widget itemBody = Material(
      child: widget.tile.setDefaults(
        defaultSelected: widget.isSelected,
        defaultTrailing: widget.items.isNotEmpty
            ? const Icon(
                Icons.arrow_forward_ios,
                size: 14,
              )
            : null,
      ),
    );

    if (isSidebarCollapse) {
      final icon =
          ((widget.tile.leading ?? const Icon(Icons.abc)) as Icon).setDefaults(
        defaultColor: widget.isSelected
            ? widget.tile.selectedColor
            : widget.tile.iconColor,
      );
      itemBody = Center(
        child: IconButton(
          icon: icon,
          onPressed: widget.tile.onTap,
        ),
      );

      if (widget.tooltip != null) {
        return Tooltip(
          message: widget.tooltip!.data ?? '',
          textStyle: widget.tooltip!.style,
          child: itemBody,
        );
      }
    }

    return itemBody;
  }
}


lib/src/section.dart
import 'package:flutter/material.dart';
import 'package:tuanis_sidebar/src/core/state.dart';
import 'package:tuanis_sidebar/tuanis_sidebar.dart';

/// Represents a section of the sidebar containing a list of items
///
/// [items] The items of this section
/// [backgroundColor] The background color of the section. Shown only if [title] is passed
/// [borderColor] Top border color, if null, no border is shown
/// [title] The text to display at the top of the section, before the items
class TuanisSidebarSection extends StatelessWidget {
  /// The items of this section
  final List<TuanisSidebarItem> items;

  /// The background color of the section. Shown only if [title] is passed
  final Color? backgroundColor;

  /// Top border color, if null, no border is shown
  final Color? borderColor;

  /// The text to display at the top of the section, before the items
  final Text? title;

  const TuanisSidebarSection({
    super.key,
    required this.items,
    this.backgroundColor,
    this.borderColor,
    this.title,
  });

  @override
  Widget build(BuildContext context) {
    final isSidebarCollapse = SidebarState.of(context)?.isCollapse ?? false;

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Container(
          width: MediaQuery.of(context).size.width,
          // padding: const EdgeInsets.symmetric(vertical: 8, horizontal: 16),
          // margin: const EdgeInsets.only(top: 16),
          decoration: BoxDecoration(
            border: borderColor != null
                ? Border(
                    top: BorderSide(color: borderColor!),
                  )
                : null,
            color: title != null && !isSidebarCollapse ? backgroundColor : null,
          ),
          child: !isSidebarCollapse ? Padding(
            padding: const EdgeInsets.all(8.0),
            child: title,
          ) : null,
        ),
        ...items,
      ],
    );
  }
}


lib/src/sidebar.dart
import 'package:flutter/material.dart';
import 'package:tuanis_sidebar/src/core/ext.dart';
import 'package:tuanis_sidebar/src/core/state.dart';
import 'package:tuanis_sidebar/src/item.dart';
import 'package:tuanis_sidebar/src/section.dart';

/// The main widget for building the sidebar
///
/// [selectedItemId] The initial selected item id
/// [selectedColor] Defines the color used for icons and text when the list tile is selected.
/// [selectedTileColor] Defines the background color of ListTile when [selected] is true.
/// [iconColor] Defines the default color for [leading] and [trailing] icons.
/// [textColor] Defines the default color for the [title] and [subtitle].
/// [focusColor] The color for the tile's [Material] when it has the input focus.
/// [hoverColor] The color for the tile's [Material] when a pointer is hovering over it.
/// [splashColor] The color of splash for the tile's [Material].
/// [tileColor] Defines the background color of ListTile when [selected] is false.
class TuanisSidebar extends StatefulWidget {
  /// Organize items into sections
  final List<TuanisSidebarSection> sections;

  ///  Control how the sections are spread within the Column
  final MainAxisAlignment sectionsAlignment;

  /// The background color of the sidebar
  final Color? backgroundColor;

  /// The initial selected item id
  final String? selectedItemId;

  /// Defines the color used for icons and text when the list tile is selected. (Applies to all items)
  final Color selectedColor;

  /// Defines the background color of ListTile when [selected] is true. (Applies to all items)
  final Color? selectedTileColor;

  /// Defines the default color for [leading] and [trailing] icons. (Applies to all items)
  final Color? iconColor;

  /// Defines the default color for the [title] and [subtitle]. (Applies to all items)
  final Color? textColor;

  /// The color for the tile's [Material] when it has the input focus. (Applies to all items)
  final Color? focusColor;

  /// The color for the tile's [Material] when a pointer is hovering over it. (Applies to all items)
  final Color? hoverColor;

  /// The color of splash for the tile's [Material]. (Applies to all items)
  final Color? splashColor;

  /// Defines the background color of ListTile when [selected] is false.  (Applies to all items)
  final Color? tileColor;

  /// The background color of the "Collapse Sidebar" item
  final Color? collapseItemBackgroundColor;

  /// If true, only icons will be shown
  final bool isCollapse;

  /// Text component for the collapse item
  final Text? collapseSidebarItemTitle;

  TuanisSidebar({
    super.key,
    required this.sections,
    this.sectionsAlignment = MainAxisAlignment.end,
    this.selectedItemId,
    this.backgroundColor,
    required this.selectedColor,
    this.selectedTileColor,
    this.iconColor,
    this.textColor,
    this.focusColor,
    this.hoverColor,
    this.splashColor,
    this.tileColor,
    this.collapseItemBackgroundColor,
    this.isCollapse = false,
    this.collapseSidebarItemTitle,
  }) {
    assert(_allItemIdsAreUnique());
  }

  @override
  State<StatefulWidget> createState() => _TuanisSidebar();

  bool _allItemIdsAreUnique() {
    final allItems = sections.map((s) {
      return [...s.items][0]; // sorry for this
    }).toList();
    return allItems.map((item) => item.id).toSet().length == allItems.length;
  }
}

class _TuanisSidebar extends State<TuanisSidebar> {
  final double _expandedSidebarWidth = 250;
  final double _collapseSidebarWidth = 70;
  double _sidebarWidth = 0;
  final double _collapseSidebarItemHeight = 60;
  String? _currentSelectedItemId;
  List<String> _expandedItemIds = [];
  bool _isCollapse = false;
  bool _futureIsCollapse = false;

  @override
  void initState() {
    super.initState();

    _currentSelectedItemId = widget.selectedItemId;
    _expandedItemIds = [];
    _isCollapse = widget.isCollapse;
    _sidebarWidth = widget.isCollapse ? _collapseSidebarWidth : _expandedSidebarWidth;
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SidebarState(
      isCollapse: _isCollapse,
      child: AnimatedContainer(
        color: widget.backgroundColor ?? Theme.of(context).scaffoldBackgroundColor,
        duration: const Duration(milliseconds: 300),
        onEnd: () {
          setState(() {
            _isCollapse = _futureIsCollapse;
          });
        },
        curve: Curves.fastOutSlowIn,
        width: _sidebarWidth,
        child: Column(
          children: [
            Expanded(child: _buildSections()),
            _buildCollapseItem(),
          ],
        ),
      ),
    );
  }

  Widget _buildSections() {
    return SizedBox(
      height: MediaQuery.of(context).size.height -
          _collapseSidebarItemHeight -
          MediaQuery.of(context).padding.top -
          MediaQuery.of(context).padding.bottom,
      child: SingleChildScrollView(
        child: ConstrainedBox(
          constraints: BoxConstraints(
            minHeight: MediaQuery.of(context).size.height -
                _collapseSidebarItemHeight -
                MediaQuery.of(context).padding.top -
                -MediaQuery.of(context).padding.bottom,
          ),
          child: Column(
            mainAxisAlignment: widget.sectionsAlignment,
            children: _getSections(),
          ),
        ),
      ),
    );
  }

  Widget _buildCollapseItem() {
    return TuanisSidebarItem(
      id: '',
      tile: ListTile(
        title: widget.collapseSidebarItemTitle ?? const Text('Collapse Sidebar'),
        leading: _isCollapse
            ? const Icon(Icons.keyboard_double_arrow_right)
            : const Icon(Icons.keyboard_double_arrow_left),
        tileColor: widget.collapseItemBackgroundColor ?? Theme.of(context).scaffoldBackgroundColor,
        onTap: () {
          setState(() {
            _futureIsCollapse = !_futureIsCollapse;
            // _isCollapse is set to false when the animation ends
            if (!_isCollapse) {
              _isCollapse = true;
            }
            if (_futureIsCollapse) {
              _sidebarWidth = _collapseSidebarWidth;
            } else {
              _sidebarWidth = _expandedSidebarWidth;
            }
          });
        },
      ),
    );
  }

  List<TuanisSidebarSection> _getSections() {
    return widget.sections
        .map(
          (e) => TuanisSidebarSection(
            backgroundColor: e.backgroundColor,
            borderColor: e.borderColor,
            title: e.title,
            items: _getItems(e.items, level: 1) as List<TuanisSidebarItem>,
          ),
        )
        .toList();
  }

  List<TuanisSidebarItem>? _getItems(List<TuanisSidebarItem> items, {required int level}) {
    if (items.isEmpty) {
      return null;
    }

    List<TuanisSidebarItem> children = [];
    for (int i = 0; i < items.length; i++) {
      children.add(_addExtraAttributesToItem(item: items[i], level: level));
      if (_expandedItemIds.contains(items[i].id)) {
        // render sub-items only if the current item is expanded
        final subItems = _getItems(items[i].items, level: level + 1);
        if (subItems != null) {
          children.addAll(subItems);
        }
      }
    }

    return children;
  }

  TuanisSidebarItem _addExtraAttributesToItem(
      {required TuanisSidebarItem item, required int level}) {
    return item.copyWith(
      extIsSelected: _currentSelectedItemId == item.id,
      extTile: item.tile.setDefaults(
        defaultContentPadding: EdgeInsets.only(left: level * 16, right: 16),
        defaultSelectedColor: widget.selectedColor,
        defaultSelectedTileColor: widget.selectedTileColor,
        defaultIconColor: widget.iconColor,
        defaultTextColor: widget.textColor,
        defaultFocusColor: widget.focusColor,
        defaultHoverColor: widget.hoverColor,
        defaultSplashColor: widget.splashColor,
        defaultTileColor: widget.tileColor,
        overrideOnTap: () {
          setState(() {
            // show the item as selected only if it is a leaf
            if (item.items.isEmpty) {
              _currentSelectedItemId = item.id;
            } else {
              // expand sub-menu
              if (_expandedItemIds.contains(item.id)) {
                _expandedItemIds.remove(item.id);
              } else {
                _expandedItemIds.add(item.id);
              }
            }

            // call user's event if not null and the item doesn't have nested items
            if (item.tile.onTap != null && item.items.isEmpty) {
              item.tile.onTap!();
            }
          });
        },
      ),
    );
  }
}


lib/src/core/ext.dart
import 'package:flutter/material.dart';

extension ColorExtensions on Color {
  Color lighten([double amount = 0.15]) {
    assert(amount >= 0 && amount <= 1);
    final hsl = HSLColor.fromColor(this);
    final hslLight =
        hsl.withLightness((hsl.lightness + amount).clamp(0.0, 1.0));
    return hslLight.toColor();
  }

  Color darken([double amount = 0.15]) {
    assert(amount >= 0 && amount <= 1);
    final hsl = HSLColor.fromColor(this);
    final hslDark = hsl.withLightness((hsl.lightness - amount).clamp(0.0, 1.0));
    return hslDark.toColor();
  }
}

extension ListTileExtensions on ListTile {
  ListTile setDefaults({
    Icon? defaultTrailing,
    EdgeInsetsGeometry? defaultContentPadding,
    GestureTapCallback? overrideOnTap,
    bool defaultSelected = false,
    Color? defaultSelectedColor,
    Color? defaultSelectedTileColor,
    Color? defaultIconColor,
    Color? defaultTextColor,
    Color? defaultFocusColor,
    Color? defaultHoverColor,
    Color? defaultSplashColor,
    Color? defaultTileColor,
  }) {
    return ListTile(
      key: key,
      leading: leading,
      title: title,
      subtitle: subtitle,
      trailing: trailing ?? defaultTrailing,
      isThreeLine: isThreeLine,
      dense: dense,
      visualDensity: visualDensity,
      shape: shape,
      style: style,
      selectedColor: selectedColor ?? defaultSelectedColor,
      iconColor: iconColor ?? defaultIconColor,
      textColor: textColor ?? defaultTextColor,
      contentPadding: contentPadding ?? defaultContentPadding,
      enabled: enabled,
      onTap: overrideOnTap ?? onTap,
      onLongPress: onLongPress,
      onFocusChange: onFocusChange,
      mouseCursor: mouseCursor,
      selected: selected || defaultSelected,
      focusColor: focusColor ?? defaultFocusColor,
      hoverColor: hoverColor ?? defaultHoverColor,
      splashColor: splashColor ?? defaultSplashColor,
      focusNode: focusNode,
      autofocus: autofocus,
      tileColor: tileColor ?? defaultTileColor,
      selectedTileColor: selectedTileColor ?? defaultSelectedTileColor,
      enableFeedback: enableFeedback,
      horizontalTitleGap: horizontalTitleGap,
      minLeadingWidth: minLeadingWidth,
      minVerticalPadding: minVerticalPadding,
    );
  }
}

extension IconExtensions on Icon {
  Icon setDefaults({Color? defaultColor, double? defaultSize}) {
    return Icon(
      icon,
      key: key,
      color: color ?? defaultColor,
      size: size ?? defaultSize,
      fill: fill,
      grade: grade,
      weight: weight,
      opticalSize: opticalSize,
      shadows: shadows,
      semanticLabel: semanticLabel,
      textDirection: textDirection,
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:tuanis_sidebar/tuanis_sidebar.dart';

void main() {
  runApp(
    const MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Example(),
    ),
  );
}

class Example extends StatelessWidget {
  const Example({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Row(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          TuanisSidebar(
            sectionsAlignment: MainAxisAlignment.spaceBetween,
            selectedItemId: 'settings',
            selectedColor: Colors.blue,
            sections: [
              const TuanisSidebarSection(
                title: Text(
                  'MAIN',
                  style: TextStyle(letterSpacing: 1.2),
                ),
                items: [
                  TuanisSidebarItem(
                    id: 'dashboard',
                    tile: ListTile(
                      leading: Icon(Icons.dashboard),
                      title: Text('Dashboard'),
                    ),
                    items: [
                      TuanisSidebarItem(
                        id: 'dashboard_1',
                        tile: ListTile(
                          leading: Icon(Icons.abc),
                          title: Text('Child of Dashboard'),
                        ),
                      )
                    ],
                  ),
                  TuanisSidebarItem(
                    id: 'settings',
                    tile: ListTile(
                      title: Text('Settings'),
                      leading: Icon(Icons.settings),
                    ),
                  ),
                ],
              ),
              TuanisSidebarSection(
                borderColor: const Color(0xffcbd5e1),
                items: [
                  TuanisSidebarItem(
                    id: 'logout',
                    tooltip: const Text(
                      'Salir de aqui',
                      style: TextStyle(
                        fontSize: 14,
                        color: Colors.white,
                      ),
                    ),
                    tile: ListTile(
                      leading: const Icon(Icons.logout),
                      title: const Text('Logout'),
                      onTap: () {
                        print('Logout');
                      },
                    ),
                  )
                ],
              ),
            ],
          ),
          Expanded(
            child: Container(
              padding: const EdgeInsets.all(50),
              color: Colors.white,
              child: const Text('App body here'),
            ),
          )
        ],
      ),
    );
  }
}
